packages/flutter/lib/src/rendering/flex.dart

```diff
@@ -12,6 +12,82 @@ import 'layer.dart';
 import 'layout_helper.dart';
 import 'object.dart';
 
+// A 2D vector that uses a [RenderFlex]'s main axis and cross axis as its first and second coordinate axes.
+// It represents the same vector as (double mainAxisExtent, double crossAxisExtent).
+extension type const _AxisSize._(Size _size) {
+  _AxisSize({ required double mainAxisExtent, required double crossAxisExtent }) : this._(Size(mainAxisExtent, crossAxisExtent));
+  _AxisSize.fromSize({ required Size size, required Axis direction }) : this._(_convert(size, direction));
+
+  static const _AxisSize empty = _AxisSize._(Size.zero);
+
+  static Size _convert(Size size, Axis direction) {
+    return switch (direction) {
+      Axis.horizontal => size,
+      Axis.vertical => size.flipped,
+    };
+  }
+  double get mainAxisExtent => _size.width;
+  double get crossAxisExtent => _size.height;
+
+  Size toSize(Axis direction) => _convert(_size, direction);
+
+  _AxisSize applyConstraints(BoxConstraints constraints, Axis direction) {
+    final BoxConstraints effectiveConstraints = switch (direction) {
+      Axis.horizontal => constraints,
+      Axis.vertical => constraints.flipped,
+    };
+    return _AxisSize._(effectiveConstraints.constrain(_size));
+  }
+
+  _AxisSize operator +(_AxisSize other) => _AxisSize._(Size(_size.width + other._size.width, math.max(_size.height, other._size.height)));
+}
+
+// The ascent and descent of a baseline-aligned child.
+//
+// Baseline-aligned children contributes to the cross axis extent of a [RenderFlex]
+// differently from chidren with other [CrossAxisAlignment]s.
+extension type const _AscentDescent._((double ascent, double descent)? ascentDescent) {
```

Comment:
The type doesn't exist at runtime I guess? Probably not much.

Reply:
It's still allocating a record at runtime. It shouldn't matter whether you allocate an instance of a class or a record? Since the extension type isn't giving us anything here my suggestion is to use the simpler concept of a regular class here.

Reply:
The `(double, double)?` pair is probably still needed even if switch to using a class, because it needs an additional bit to represent the "no baselines" state.